src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2182,516 @@ def terminate(self):
+
+        if self._stage is None:
+            raise ValueError("LinkedHeightActuator needs a stage dependency.")
+        if self._lensz is None:
+            raise ValueError("LinkedHeightActuator needs a lensz dependency.")
+
+        axes_def = {}
+        for an in self._stage.axes.keys():
+            axes_def[an] = copy.deepcopy(self._stage.axes[an])
+            axes_def[an].canUpdate = False
+
+        model.Actuator.__init__(self, name, role, axes=axes_def, dependencies=dependencies,
+                                **kwargs)
+        self._metadata[model.MD_HW_NAME] = "LinkedHeightActuator"
+
+        # will take care of executing axis moves asynchronously
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+
+        # position will be updated directly from the underlying stage
+        self._position = {}
+        self.position = model.VigilantAttribute({}, unit="m", readonly=True)
+        self._updatePosition()
+
+        self.referenced = model.VigilantAttribute({}, readonly=True)
+        # listen to reference changes from the underlying stage
+        logging.debug("Subscribing to reference of dependency %s", self._stage.name)
+        self._stage.referenced.subscribe(self._ondepStageReferenced)
+        self._ondepStageReferenced(self._stage.referenced.value)
+
+        self._focus = None
+        # Create a linked height focus child object to control the movement of the lens stage Z axis
+        if "focus" in children:
+            ckwargs = children["focus"]
+            self._focus = LinkedHeightFocus(parent=self, dependencies={"lensz": self._lensz}, **ckwargs)
+            self.children.value.add(self._focus)
+        else:
+            raise ValueError("Focus should be in actuator's children.")
+
+        # Add speed if it's in stage
+        if model.hasVA(self._stage, "speed"):
+            self.speed = self._stage.speed
+
+    @roattribute
+    def executor(self):
+        """ Expose executor as property to be shared by the child focus object"""
+        return self._executor
+
+    def updateMetadata(self, md):
+        self._metadata.update(md)
+
+    def _updatePosition(self):
+        """
+        update the position VA from the underlying dependency
+        """
+        self._position = self._stage.position.value
+        self.position._set_value(self._position, force_write=True)
+
+    @isasync
+    def moveAbs(self, pos):
+        """
+        Move the sample stage to the defined position in m for each axis given, and adjust focus if needed. This is an
+        asynchronous method.
+        pos dict(string-> float): name of the axis and new position in m
+        returns (Future): object to control the move request
+        """
+        if not pos:
+            pos = self.position.value
+        self._checkMoveAbs(pos)
+
+        f = self._executor.submit(self._doMoveAbs, pos)
```

Comment:
All moves need to be cancellable. When cancelling happens, the current move should stop, and if there were still some (internal) moves planned, they shouldn't be executed.

Reply:
Thanks for the wholesome amazing review! I made most of the changes but I'm having trouble with the movement cancellation. The movement stops when it's cancelled but the `_was_stopped` attribute always return `False`. Can you please just have a look if I'm indeed on the right way on the cancellation, so I can investigate this issue further.. :)